### PR TITLE
Do not modify query in Search::count

### PR DESCRIPTION
**Pull request by XWB**
Related to https://github.com/ruflin/Elastica/issues/1261

### CODE REVIEW

**Comment by z38** [comment]
It might be better to clone the query and keep the removed line. With the current fix, the first 10(?) results get transmitted over the wire.

**Comment by ruflin** [comment]
@XWB Thanks for the PR. I think I understand the problem now better. It becomes a problem if count is used as the query itself is modified. Shouldn't we use a count query (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-count.html) here if fullResult is not set? I'm curious why we introduce fullResult here anyways.

Currently at Elastic{ON} so will not have time this week to dig into this.

**Comment by ruflin** [comment]
@XWB Had a quick look at it, and it seems we should use a _count query if fullResult is false. Or do I miss something here?

**Comment by Tobion** [comment]
This is not the correct solution. Setting the size to 0 is an important optimization as also pointed out by @z38 and the main advantage of using the shortcut `count()` at all.
But I agree the query should not be modified. Please clone the query before setting the size and please also add a test for this.

**Comment by Tobion** [comment]
@XWB do you have time to finish this?

**Comment by XWB** [comment]
@Tobion Will do.

**Comment by ruflin** [comment]
@XWB Can you also update the CHANGELOG.md?

**Comment by XWB** [comment]
@ruflin @Tobion Done.

File: lib/Elastica/Search.php [hunk 1]
```diff
@@ -475,7 +475,7 @@ public function count($query = '', $fullResult = false)
     {
         $this->setOptionsAndQuery(null, $query);
 
-        $query = $this->getQuery();
+        $query = clone $this->getQuery();
```
Comment (ruflin) [inline lib/Elastica/Search.php]: Could you add a quick note (docs comment) on what it is cloned? Just to make sure nobody will remove it in the future ...

**Comment by XWB** [comment]
@ruflin Done.

**Comment by ruflin** [comment]
@Tobion Do you want to have a look before merging?

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,7 @@ All notable changes to this project will be documented in this file based on the
 - Fix reading bool index settings like `\Elastica\Index\Settings::getBlocksWrite`. Elasticsearch returns all settings as strings and does not normalize bool values.
   The getters now return the right bool value for whichever string representation is used like 'true', '1', 'on', 'yes'.
 - Fix for QueryBuilder version check `\Elastica\QueryBuilder\Version\Version240.php` added all new query types to queries array.
+- Do not modify the original query in `\Search::count`.
```
Comment (Tobion) [inline CHANGELOG.md]: `\Elastica\Search`

File: test/Elastica/SearchTest.php [hunk 3]
```diff
@@ -578,4 +578,29 @@ public function testIgnoreUnavailableOption()
...
+    /**
+     * @group functional
+     */
+    public function testQuerySizeAfterCount()
+    {
+        $index = $this->_createIndex();
+        $search = new Search($index->getClient());
+        $type = $index->getType('test');
+
+        $doc = new Document(1, ['id' => 1, 'username' => 'ruflin']);
```
Comment (Tobion) [inline test/Elastica/SearchTest.php]: No need to add a document since it's not relevant for this test.

**Comment by XWB** [comment]
@Tobion Changes are underway.

**Comment by XWB** [comment]
And done.

File: test/Elastica/SearchTest.php [hunk 4]
```diff
@@ -578,4 +578,26 @@ public function testIgnoreUnavailableOption()
         $results = $search->search($query, [Search::OPTION_SEARCH_IGNORE_UNAVAILABLE => true]);
         $this->assertInstanceOf(ResultSet::class, $results);
     }
+
+    /**
+     * @group functional
+     */
+    public function testQuerySizeAfterCount()
+    {
+        $index = $this->_createIndex();
```
Comment (Tobion) [inline test/Elastica/SearchTest.php]: you also don't need an index or a type for this test.

**Comment by XWB** [comment]
@Tobion I simplified the test.

**Pull request merged**